src/main/java/io/neonbee/internal/codec/DataExceptionMessageCodec.java

```diff
@@ -0,0 +1,117 @@
+package io.neonbee.internal.codec;
+
+import java.util.Map;
+
+import com.google.common.base.Strings;
+
+import io.neonbee.data.DataException;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.eventbus.MessageCodec;
+import io.vertx.core.json.JsonArray;
+import io.vertx.core.json.JsonObject;
+
+public class DataExceptionMessageCodec implements MessageCodec<DataException, DataException> {
+
+    @Override
+    public void encodeToWire(Buffer buffer, DataException body) {
+        Map<String, Object> failureBody = body.failureDetail();
+        buffer.appendInt(body.failureCode());
+        if (!Strings.isNullOrEmpty(body.getMessage())) {
+            appendObjectToBuffer(body.getMessage(), buffer);
+        }
+        if (!body.failureDetail().isEmpty()) {
+            failureBody.keySet().forEach(key -> {
+                appendObjectToBuffer(key, buffer);
+                appendObjectToBuffer(failureBody.get(key), buffer);
+            });
+        }
```

Comment:
How about something along those lines: First I think body should be renamed `exception` or `ex` to make it more clear what is being encoded. Then:

```java
buffer.appendInt(ex.failureCode())

String msg = ex.getMessage();
if (msg != null) {
  byte[] strBytes = s.getBytes(CharsetUtil.UTF_8);
  buffer.appendInt(strBytes.length);
  buffer.appendBytes(strBytes);
} else {
  buffer.appendInt(-1);
}

Buffer encDetail = new JsonObject(ex.failureDetail()).toBuffer(); // we can also put this in the JavaDoc of the DataException, that the failureMap must be compatible to the JsonObject constructor (you can link it with {@link JsonObject(Map)} so it is clear to the end-user
buffer.appendInt(encDetail.length());
buffer.appendBuffer(encDetail);
```

Then the decoding can be also much simplified! 👍